### PR TITLE
fix(key-auth): keep query params order if hide_credentials is true

### PR DESCRIPTION
**Pull request by battlebyte**
When hide_credentials=true, query parameters should not be sorted alphabetically. This alters the original request and can have unintended consequences for upstream services. One example is authentication with the pseudo header (request-target) as described in https://datatracker.ietf.org/doc/html/draft-cavage-http-signatures#section-2.3. Since the order is not preserved, the signature does not match and the upstream rejects the request.

* query parameters order is not altered when hide_credentials=true
* add two new test cases in spec/03-plugins/09-key-auth

### Checklist

- [X] The Pull Request has tests
- [X] A changelog file has been created under `changelog/unreleased/kong` or `skip-changelog` label added on PR if changelog is unnecessary. [README.md](https://github.com/Kong/gateway-changelog/README.md)
- [ ] There is a user-facing docs PR against https://github.com/Kong/docs.konghq.com - PUT DOCS PR HERE

---
- To see the specific tasks where the Asana app for GitHub is being used, see below:
  - https://app.asana.com/0/0/1206863362995171

### CODE REVIEW

**Review by gszr — CHANGES_REQUESTED** [review]
@battlebyte A few thoughts...

- The [RFC2616 (HTTP 1.1) Section 4.2](https://datatracker.ietf.org/doc/html/rfc2616#section-4.2) says:
> The order in which header fields with differing field names are
   received is not significant. However, it is "good practice" to send
   general-header fields first, followed by request-header or response-
   header fields, and ending with the entity-header fields.

Thus, Kong's behavior is not deviating from the RFC.

The draft you linked also reinforces the fact that there may be alterations:

>  A signed HTTP message needs to be tolerant of some trivial
   alterations during transmission as it goes through gateways, proxies,
   and other entities

In the same section, the draft also says:
> In order to generate the string that is signed with a key, the client
   MUST use the values of each HTTP header field in the `headers`
   Signature Parameter, in the order they appear in the `headers`
   Signature Parameter.

The verification also `MUST` follow the same process to recreate the signature (https://datatracker.ietf.org/doc/html/draft-cavage-http-signatures#section-2.5).

Thus, I see this as a non-conforming implementation either on the client signing or the server verifying the signature.

**Comment by battlebyte** [comment]
I agree with your points @gszr , although those are in relation to the headers. The order of the headers is not important because, as you point out, the client MUST use the values of each HTTP header field in the headers Signature Parameter, in the order they appear in the headers Signature Parameter.

The point here is that one of those headers is the (request-target) pseudo-header. This header includes the path and the query string, that is the reason why the order of the query string arguments should not be altered. See an example in slide 6 here https://revenue-ie.github.io/paye-employers-documentation/PIT3/examples/REST_Request_Authentication_V1.pdf. 

I hope this clarifies.

**Comment by battlebyte** [comment]
If you see the example in https://datatracker.ietf.org/doc/html/draft-cavage-http-signatures#section-3.1.3, the headers that will be part of the signature are:

`headers="(request-target) (created) host digest content-length"`

Those headers can be received in any order, that is not a problem because the upstream knows how they should be ordered to calculate the signature.

The key point, is that the `(request-target)` header contains the HTTP method and the request path, eventually including the query parameters. For example:

`(request-target): get /this/is/a/sample/path?zparam=1&tparam=2&aparam=3`

With the current implementation of `key-auth` plugin, if `hide_credentials=true`, the order of the query params are altered and placed in alphabetical order. So the upstream will receive the request as:

`/this/is/a/sample/path?aparam=3&tparam=2&zparam=1`

And the signature calculated by the upstream will not match the signature provided by the client.

**Review by battlebyte — COMMENTED** [review]
See my comments in the main conversation.

**Comment by gszr** [comment]
@battlebyte I have taken the liberty to rebase and make a couple of changes to the code:

- Use `ngx.re.gsub` with the `oj` flags (explanation [here](https://gist.github.com/gszr/95f78639dc3c4545d199becf9d789d68) for anyone interested) as commonly done in our code base
- Remove some formatting issues

I will also request @bungle to take a look at this one.

Also, the oauth2 plugin removes the header in the same way; should this fix also be applied there?

**Review by gszr — DISMISSED** [review]

**Comment by battlebyte** [comment]
Thanks for the review and changes @gszr .
Regarding oauth2, I think the same fix should be applied there to avoid altering the query params ordering.

**Comment by gszr** [comment]
> Thanks for the review and changes @gszr . Regarding oauth2, I think the same fix should be applied there to avoid altering the query params ordering.

Agreed; once this one is merged we can submit a separate PR.

**Comment by bungle** [comment]
Such a small PR which on a surface looks good and perhaps better than what we have, just raises a lot of small questions, for example:

Can querystring param names be urlencoded, e.g. `key=` -> `%6B%65%79=` or `k%65y=`, this will complicate the processing. Or is this somehow handled already (let's figure this out).

File: kong/plugins/key-auth/handler.lua
Comment (gszr) [inline kong/plugins/key-auth/handler.lua:110]: @battlebyte - chatting with @bungle, we questioned what would happen if the query arg is url-encoded. Can you think about this for a bit and verify?

Comment (battlebyte) [reply kong/plugins/key-auth/handler.lua:110]: That is a good point and I thought about that. According to the plugin documentation, 

> The key names may only contain [a-z], [A-Z], [0-9], [_] underscore, and [-] hyphen.

which looks to me those were carefully chosen to avoid url-encoding. I checked the url encoding functions of several programming languages and they do not encode those characters. For example, those won't be encoded in

> [ngx.escape_uri](https://github.com/openresty/lua-nginx-module?tab=readme-ov-file#ngxescape_uri)

But certainly, it still could be that it is url-encoded. I could use 

> [ngx.unescape_uri](https://github.com/openresty/lua-nginx-module?tab=readme-ov-file#ngxunescape_uri)

 to make sure we are dealing with a url-decoded key name before applying the substitution. 

 What do you think?

Comment (gszr) [reply kong/plugins/key-auth/handler.lua:110]: What about the separator `&` - that we are using in the regex?

Comment (battlebyte) [reply kong/plugins/key-auth/handler.lua:110]: You mean the `key` configured is not the `key` used in the path? Can you provide an example?

Comment (gruceo) [reply kong/plugins/key-auth/handler.lua:110]: `key_name` = `apikey`
but `apikey` is not found on the request's querystring

`key_name` = `apikey`
request's querystring contains typos: `aapikey`, `apikeyy`, or contains unfinished keys: `key` or `api`

`key_name` = `apikey`
request's querystring is malformed:
`"/request?apikey==value1"`
`"/request?apikey=kong&&a=value3",`

File: kong/plugins/key-auth/handler.lua
Comment (gruceo) [inline kong/plugins/key-auth/handler.lua:112]: any error handling we can add here? ngx.re.sub can return errors: https://github.com/openresty/lua-nginx-module/?tab=readme-ov-file#ngxresub

or we can add it in the calling function itself

```
local new_query, err = remove_query_key(raw_query, name)
if err then
  ..
end
```

Comment (gszr) [reply kong/plugins/key-auth/handler.lua:112]: @gruceo - i believe only malformed regex / replace args will reasonably trigger errors here. what are some other errors?

File: kong/plugins/key-auth/handler.lua
Comment (gszr) [reply kong/plugins/key-auth/handler.lua:110]: Any concerns here @bungle?

Comment (battlebyte) [reply kong/plugins/key-auth/handler.lua:111]: Latest commit uses ngx.unescape_uri to remove the key only on uri decoded query strings.

Comment (battlebyte) [reply kong/plugins/key-auth/handler.lua:111]: Those cases are already managed in the code. If the exact key is not found, a `401 Unauthorized` with message `"No API key found in request"` is returned.

File: kong/plugins/key-auth/handler.lua
Comment (battlebyte) [reply kong/plugins/key-auth/handler.lua:119]: The latest commit handles the error.

File: kong/plugins/key-auth/handler.lua
Comment (gszr) [inline kong/plugins/key-auth/handler.lua:204]: please revert these unrelated changes

File: kong/plugins/key-auth/handler.lua
Comment (gszr) [inline kong/plugins/key-auth/handler.lua:225]: ditto

File: kong/plugins/key-auth/handler.lua
Comment (gszr) [inline kong/plugins/key-auth/handler.lua:240]: ditto

File: spec/03-plugins/09-key-auth/02-access_spec.lua
Comment (gszr) [inline spec/03-plugins/09-key-auth/02-access_spec.lua:744]: this test case does not exercise the case where the separator `&` is also encoded; let's extend it to include that

File: kong/plugins/key-auth/handler.lua
Comment (gszr) [reply kong/plugins/key-auth/handler.lua:119]: @battlebyte the second `ngx_re_gsub` is not handling the error

File: kong/plugins/key-auth/handler.lua
Comment (gszr) [inline kong/plugins/key-auth/handler.lua:164]: If an error occurs in the regex gsub, `remove_query_key` will return `nil`, and this call to `set_raw_query` will now throw an error. Let's change `remove_query_key` to return `nil, err` and log the error from here instead. 
Also, I wonder if now we should respond with a 500 in case the query key clearing cannot be done - it can also be seen as a security issue if the plugin config has `hide_credentials` and it has no effect.

**Review by gszr — CHANGES_REQUESTED** [review]

File: kong/plugins/key-auth/handler.lua
Comment (battlebyte) [reply kong/plugins/key-auth/handler.lua]: Fixed in the latest commit. I think returning 500 makes sense but I'm waiting on the final decision you guys make.

**Review by gruceo — APPROVED** [review]

**Review by gszr — DISMISSED** [review]

File: kong/plugins/key-auth/handler.lua
Comment (gszr) [reply kong/plugins/key-auth/handler.lua]: This is a new error condition; IMO it makes sense to respond with a 500 in that case. Otherwise, the token will be leaked to the upstream service, even though the expectation is it was removed.

**Comment by bungle** [comment]
Update:

The Ada bindings look quite fine:
https://github.com/bungle/lua-resty-ada

And the Ada library is proposed to be added here:
https://github.com/Kong/kong/pull/13120

**Comment by battlebyte** [comment]
Thanks @bungle , that makes sense. Your proposal is a much better solution. Thanks for taking the time!

**Comment by battlebyte** [comment]
Hi @bungle , what are the next steps to close this topic?

**Pull request closed without merging**
